adapters/repos/db/queue/scheduler.go

```diff
@@ -0,0 +1,490 @@
+	chans := make([]chan Batch, s.Workers)
+
+	for i := 0; i < s.Workers; i++ {
+		worker, ch := NewWorker(s.Logger, 5*time.Second)
+		chans[i] = ch
+
+		s.wg.Add(1)
+		f := func() {
+			defer s.wg.Done()
+
+			worker.Run(s.ctx)
+		}
+		enterrors.GoWrapper(f, s.Logger)
+	}
+
+	s.chans = chans
+
+	// run scheduler goroutine
+	s.wg.Add(1)
+	f := func() {
+		defer s.wg.Done()
+
+		s.runScheduler()
+	}
+	enterrors.GoWrapper(f, s.Logger)
+}
+
+func (s *Scheduler) Close() error {
+	if s == nil || s.ctx == nil {
+		// scheduler not initialized. No op.
+		return nil
+	}
+
+	// check if the scheduler is already closed
+	if s.ctx.Err() != nil {
+		return nil
+	}
+
+	// stop scheduling
+	s.cancelFn()
+
+	// wait for the workers to finish processing tasks
+	s.activeTasks.Wait()
+
+	// wait for the spawned goroutines to stop
+	s.wg.Wait()
+
+	// close the channels
+	for _, ch := range s.chans {
+		close(ch)
+	}
+
+	s.Logger.Debug("scheduler closed")
+
+	if s.OnClose != nil {
+		s.OnClose()
+	}
+
+	return nil
+}
+
+func (s *Scheduler) PauseQueue(id string) {
+	q := s.getQueue(id)
+	if q == nil {
+		return
+	}
+
+	q.m.Lock()
+	q.paused = true
+	q.m.Unlock()
+
+	s.Logger.WithField("id", id).Debug("queue paused")
+}
+
+func (s *Scheduler) ResumeQueue(id string) {
+	q := s.getQueue(id)
+	if q == nil {
+		return
+	}
+
+	q.m.Lock()
+	q.paused = false
+	q.m.Unlock()
+
+	s.Logger.WithField("id", id).Debug("queue resumed")
+}
+
+func (s *Scheduler) Wait(id string) {
+	q := s.getQueue(id)
+	if q == nil {
+		return
+	}
+
+	q.scheduled.Wait()
+	q.activeTasks.Wait()
+}
+
+func (s *Scheduler) WaitAll() {
+	s.activeTasks.Wait()
+}
+
+func (s *Scheduler) getQueue(id string) *queueState {
+	s.queues.Lock()
+	defer s.queues.Unlock()
+
+	return s.queues.m[id]
+}
+
+func (s *Scheduler) runScheduler() {
+	t := time.NewTicker(s.ScheduleInterval)
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			// stop the ticker
+			t.Stop()
+			return
+		case <-t.C:
+			s.schedule()
+		case ch := <-s.triggerCh:
+			s.scheduleQueues()
+			close(ch)
+		}
+	}
+}
+
+// Manually schedule the queues.
+func (s *Scheduler) Schedule(ctx context.Context) {
+	ch := make(chan struct{})
+	select {
+	case s.triggerCh <- ch:
+		select {
+		case <-ch:
+		case <-ctx.Done():
+		}
+	default:
+	}
+}
+
+func (s *Scheduler) schedule() {
+	// as long as there are tasks to schedule, keep running
+	// in a tight loop
+	for {
+		if s.ctx.Err() != nil {
+			return
+		}
+
+		if nothingScheduled := s.scheduleQueues(); nothingScheduled {
+			return
+		}
+	}
+}
+
+func (s *Scheduler) scheduleQueues() (nothingScheduled bool) {
+	// loop over the queues in random order
+	s.queues.Lock()
+	ids := make([]string, 0, len(s.queues.m))
+	for id := range s.queues.m {
+		ids = append(ids, id)
+	}
+	s.queues.Unlock()
+
+	nothingScheduled = true
+
+	for _, id := range ids {
+		if s.ctx.Err() != nil {
+			return
+		}
+
+		q := s.getQueue(id)
+		if q == nil {
+			continue
+		}
+
+		// mark it as scheduled
+		q.MarkAsScheduled()
```

Comment:
in this point q is not nil but there is no warranty it's still in the list of queues

Reply:
Exactly, and it's totally fine as this can also happen lower in the chain